### PR TITLE
feat: add parameter `index`

### PR DESCRIPTION
**Pull request by 111xumengze**

### CODE REVIEW

**Comment by rxaviers** [comment]
Thank you for your PR.

Having the `index` as you suggest is more uniform with other JS API such as [Array.prototype.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) and looks good to me.

I am open to merge and release it in a next major bump (because it's an API breaking change) after having some small adjustments (review comments to follow)

File: lib/es7.js [hunk 1]
```diff
@@ -13,8 +13,8 @@ async function asyncPool(poolLimit, array, iteratorFn) {
   }
   const ret = [];
   const executing = [];
-  for (const item of array) {
-    const p = Promise.resolve().then(() => iteratorFn(item, array));
+  for (const [idx, item] of array.entries()) {
+    const p = Promise.resolve().then(() => iteratorFn(item, idx, array));
```
Comment (rxaviers) [inline lib/es7.js]: Let's use the full word?

```suggestion
  for (const [index, item] of array.entries()) {
    const p = Promise.resolve().then(() => iteratorFn(item, index, array));
```

File: lib/es6.js [hunk 2]
```diff
@@ -22,8 +22,9 @@ function asyncPool(poolLimit, array, iteratorFn) {
     if (i === array.length) {
       return Promise.resolve();
     }
+    const currentIdx = i;
```
Comment (rxaviers) [inline lib/es6.js:25]: Let's use the same variable name (for consistency)
```suggestion
    const index = i;
```

File: lib/es6.js [hunk 3]
```diff
@@ -22,8 +22,9 @@ function asyncPool(poolLimit, array, iteratorFn) {
...
     const item = array[i++];
-    const p = Promise.resolve().then(() => iteratorFn(item, array));
+    const p = Promise.resolve().then(() => iteratorFn(item, currentIdx, array));
```
Comment (rxaviers) [inline lib/es6.js:27]:
```suggestion
    const p = Promise.resolve().then(() => iteratorFn(item, index, array));
```

File: lib/es6.js (hunk 2 above)
Comment (rxaviers) [inline lib/es6.js]: Can't we just use `i` below?
```suggestion
```

File: lib/es6.js (hunk 3 above)
Comment (rxaviers) [inline lib/es6.js]:
```suggestion
    const p = Promise.resolve().then(() => iteratorFn(item, i, array));
```

File: README.md [hunk 4]
```diff
@@ -71,7 +71,7 @@ Input array.
 
 #### iteratorFn
 
-Iterator function that takes two arguments (array item and the array itself). The iterator function should either return a promise or be an async function.
+Iterator function that takes two arguments (array item, array index and the array itself). The iterator function should either return a promise or be an async function.
```
Comment (rxaviers) [inline README.md]: Let's borrow some words from MDN Array/forEach https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach
```suggestion
Iterator function that takes three arguments (element value, element index, and the array object being traversed). The iterator function should either return a promise or be an async function.
```

**Comment by rxaviers** [comment]
Waiting on your input to proceed... Thanks

**Comment by rxaviers** [comment]
I want to point out it's already possible to make the index available to the iterator. Taking the example from README.md, making index available looks like this:

```diff
- const timeout = (i) =>
+ const timeout = (k, i) =>
   new Promise((resolve) => {
-    setTimeout(() => resolve(i), i);
+    setTimeout(() => resolve([k, i]), i);
   });
 await asyncPool(
   2,
-  [1000, 5000, 3000, 2000],
+  [1000, 5000, 3000, 2000].entries(),
   timeout
 );
```

Let's add an example (ideally simpler than the one I put above) to README to cover this use case?

File: lib/es6.js (hunk 3 above)
Comment (111xumengze) [reply lib/es6.js]: > 
Thank you for your reviewing.
You know, there is a `const item = array[i++]` in front.

**Comment by rxaviers** [comment]
Closing this PR, specially due to https://github.com/rxaviers/async-pool/pull/29#issuecomment-963261688

**Pull request closed without merging**